consensus/src/next_leader.rs

```diff
@@ -87,63 +96,76 @@ where
 
         let lock = self.vote_collection_chan.lock().await;
         while let Ok(msg) = lock.recv().await {
-            if msg.view_number() != self.cur_view {
+            if Into::<ConsensusMessage<_, _, _>>::into(msg.clone()).view_number() != self.cur_view {
                 continue;
             }
             match msg {
-                ConsensusMessage::TimedOut(t) => {
-                    qcs.insert(t.justify_qc);
-                }
-                ConsensusMessage::Vote(vote) => {
-                    // if the signature on the vote is invalid,
-                    // assume it's sent by byzantine node
-                    // and ignore
-
-                    if !self.api.is_valid_signature(
-                        &vote.signature.0,
-                        &vote.signature.1,
-                        vote.leaf_commitment,
-                        vote.current_view,
-                        // Ignoring deserialization errors below since we are getting rid of it soon
-                        Unchecked(vote.vote_token.clone()),
-                    ) {
-                        continue;
-                    }
+                ProcessedConsensusMessage::Vote(vote_message, sender) => {
+                    match vote_message {
+                        Vote::Yes(vote) => {
+                            if vote.signature.0
+                                != <TYPES::SignatureKey as SignatureKey>::to_bytes(&sender)
+                            {
+                                continue;
+                            }
+
+                            // if the signature on the vote is invalid,
+                            // assume it's sent by byzantine node
+                            // and ignore
+
+                            if !self.api.is_valid_vote(
+                                &vote.signature.0,
+                                &vote.signature.1,
+                                VoteData::Yes(vote.leaf_commitment),
```

Comment:
Unlike `block.commit()` that we can directly use to validate the DA vote, here we don't have something like `leaf.commit()` that we know is true. But if we use `vote.leaf_commitment` for the validation, we need a way to ensure it's correct/consistent among honest nodes.

We can probably change `stake_casted` to a map from leaf commitment to count, and we construct the QC only when the count of a commitment meets the threshold.